devices/src/virtio/queue.rs

```diff
@@ -105,12 +106,30 @@ impl<'a> DescriptorChain<'a> {
         }
     }
 
+    // Checks if current descriptor head is valid as follows:
+    // - address range is inside a mapped memory region
+    // - next descriptor from the chain fits into queue
+    // - implicit check of len gt 0
     fn is_valid(&self) -> bool {
-        !(self
-            .mem
-            .checked_offset(self.addr, self.len as usize)
-            .is_none()
-            || (self.has_next() && self.next >= self.queue_size))
+        // Check if address length overflows from usize, and get upper limit
+        let upper_limit_opt = self.addr.checked_add(self.len as usize);
+        if upper_limit_opt.is_none() {
+            return false;
+        }
+
+        // Check if last address, below upper_limit, overflows usize.
+        let last_addr_opt = upper_limit_opt.unwrap().checked_sub(1);
+        if last_addr_opt.is_none() {
+            return false;
+        }
```

Comment:
I don't really understand this one either. `x.checked_sub(1)` will only fail for `x == 0`, so what is that you are checking here?